test/pool_test.dart

```diff
@@ -105,6 +105,68 @@ void main() {
     });
   });
 
+  group("streamWithResource()", () {
+    test("The Stream holds a resource until it is completed.", () {
+      new FakeAsync().run((async) {
+        StreamController<int> streamController = new StreamController<int>();
+        var pool = new Pool(1);
+        pool.streamWithResource(() => streamController.stream);
+
+        bool nextTaskCompleted = false;
+        pool.withResource(() {
+          nextTaskCompleted = true;
+        });
+
+        new Future.delayed(new Duration(microseconds: 1), () {
+          expect(nextTaskCompleted, isFalse);
+          streamController.close();
+          return new Future.delayed(new Duration(microseconds: 1));
+        }).then((_) {
+          expect(nextTaskCompleted, isTrue);
+        });
+
+        async.elapse(new Duration(seconds: 1));
+      });
+    });
+
+    test("Cancelling the Stream propagates and releases the resource.", () {
+      new FakeAsync().run((async) {
+        StreamController<int> streamController = new StreamController<int>();
+        var pool = new Pool(1);
+        StreamSubscription<int> subscription = pool
+            .streamWithResource(() => streamController.stream)
+            .listen((_) {});
+
+        bool nextTaskCompleted = false;
+        pool.withResource(() {
+          nextTaskCompleted = true;
+        });
+
+        new Future.delayed(new Duration(microseconds: 1), () {
+          expect(nextTaskCompleted, isFalse);
+          subscription.cancel();
+          return new Future.delayed(new Duration(microseconds: 1));
+        }).then((_) {
+          expect(nextTaskCompleted, isTrue);
+        });
+
+        async.elapse(new Duration(seconds: 1));
+      });
+    });
+
+    test("Simple Stream example.", () {
+      new FakeAsync().run((async) {
+        var pool = new Pool(1);
+        pool.streamWithResource(() => new Stream.fromIterable([0, 1, 2]))
+            .toList()
+            .then((list) {
+              expect(list, [0, 1, 2]);
+            });
+        async.elapse(new Duration(seconds: 1));
+      });
+    });
```

Comment:
Also test that the resource is released if the callback returns `null`, and that the callback isn't run until a resource is available.